### PR TITLE
nautilus: mgr: Balancer fixes

### PR DESCRIPTION
**Pull request by dzafman**
Backports:
#31542
#31774
#31990
#32247

Parent trackers
https://tracker.ceph.com/issues/42756
https://tracker.ceph.com/issues/42718
https://tracker.ceph.com/issues/43084 - no backport tracker
https://tracker.ceph.com/issues/43307
https://tracker.ceph.com/issues/43312

Backport trackers
https://tracker.ceph.com/issues/42797
https://tracker.ceph.com/issues/43092
https://tracker.ceph.com/issues/43529
https://tracker.ceph.com/issues/43530

## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by smithfarm** [comment]
@dzafman Does this need https://github.com/ceph/ceph/pull/31944 as well?

**Comment by dzafman** [comment]
@smithfarm This now includes #31990 which replaced #31944 which was merge with a different change.  The tracker has been corrected too.

**Comment by dzafman** [comment]
Testing passed:
http://pulpito.ceph.com/dzafman-2019-12-05_19:53:40-rados-wip-zafman-testing-nautilus-distro-basic-smithi/

File: PendingReleaseNotes [hunk 1]
```diff
@@ -116,3 +116,13 @@
 
 * New OSD daemon command dump_scrub_reservations which reveals the
   scrub reservations that are held for local (primary) and remote (replica) PGs.
+
```
Comment (smithfarm) [inline PendingReleaseNotes]: Please put a heading here:

```
14.2.6
--------
```

**Comment by dzafman** [comment]
Waiting for #32247 to merge and then added here before merging

**Comment by yuriw** [comment]
[wip-yuri3-testing-2020-01-08-2028-nautilus](https://trello.com/c/5X40EXaX/796-wip-yuri3-testing-2020-01-08-2028-nautilus)

**Comment by yuriw** [comment]
[wip-yuri3-testing-2020-01-08-2028-nautilus](https://trello.com/c/5X40EXaX/796-wip-yuri3-testing-2020-01-08-2028-nautilus)

**Comment by neha-ojha** [comment]
@yuriw I think @dzafman is adding more commits to this PR, let's hold off on testing this.

**Comment by dzafman** [comment]
@neha-ojha I thought it would be cool to try GitHub fixing a simple conflict with PendingReleaseNotes.  Now it did whatever it did and there is an extra merge to Nautilus.

**Comment by dzafman** [comment]
@yuriw @neha-ojha I just pushed a new branch with additional change.  Let's make sure make check passes, because I was unable to run that on my build machine.

**Comment by dzafman** [comment]
@yuriw @neha-ojha There were 4 seastar failures which shouldn't be related to my changes.

	167 - unittest_seastar_buffer (Failed)
	169 - unittest_seastar_messenger (Failed)
	170 - unittest_seastar_thread_pool (Failed)
	171 - unittest_seastar_perfcounters (Failed)

**Comment by dzafman** [comment]
@yuriw Maybe my conflict resolution will break your testing....I'm going to force push and remove it.

**Comment by tchaikov** [comment]
probably we can remove `unittest_seastar_*` tests from nautilus.

**Comment by smithfarm** [comment]
Fails to compile due to an issue with boost ("begin is not a member of 'boost'"):

```
In file included from /opt/ceph/include/boost/coroutine/coroutine.hpp:10,
                 from /opt/ceph/include/boost/coroutine/all.hpp:11,
                 from /opt/ceph/include/boost/asio/spawn.hpp:19,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests/src/common/async/yield_context.h:32,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_pubsub_push.h:9,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_pubsub_push.cc:4:
/opt/ceph/include/boost/coroutine/asymmetric_coroutine.hpp: In function 'typename boost::coroutines::pull_coroutine<R>::iterator boost::coroutines::begin(boost::coroutines::pull_coroutine<Arg>&)':
/opt/ceph/include/boost/coroutine/asymmetric_coroutine.hpp:2364:17: error: 'begin' is not a member of 'boost'
 2364 | { return boost::begin( c); }
      |                 ^~~~~
```

**Comment by smithfarm** [comment]
jenkins test make check

**Comment by smithfarm** [comment]
jenkins test make check

**Comment by smithfarm** [comment]
@dzafman Now that https://github.com/ceph/ceph/pull/32658 has been merged, I guess the last commit can be dropped and that will fix the conflict?

**Comment by dzafman** [comment]
@smithfarm I removed the redundant conflicting commit, so it can be merged.

**Pull request merged**